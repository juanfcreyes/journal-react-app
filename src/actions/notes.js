import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotesDB } from "../helpers/loadNotesDB";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        try {
            const doc = await db.collection(`${auth.uid}/journal/notes`).add(newNote);
            dispatch(setActiveNote(doc.id, newNote));
            dispatch(addNewNote({ id: doc.id, ...newNote}));
        } catch (error) {
            console.error(error);
        }
    
    }
};

export const startLoadNotes = () => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        try {
            const notes = await loadNotesDB(auth.uid);
            dispatch(loadNotes(notes));
        } catch (error) {
            dispatch(loadNotes([]));
            console.log(error);
        }
    }
};

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { auth } = getState();
        const noteFS = {...note};
        delete noteFS.id;
        try {
            await db.doc(`${auth.uid}/journal/notes/${note.id}`).update(noteFS);     
        } catch (error) {
            console.log('startSaveNote error', error)
        }
        dispatch(refreshNote(note.id, noteFS));
        dispatch(setActiveNote(note.id, noteFS))
        Swal.fire('Save', note.title, 'success');
    }
}

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: note } = getState().notes;
        showUploadAlert();
        try {
            const fileUrl = await fileUpload(file);
            note.url = fileUrl;        
        } catch (error) {
            
        }
        dispatch(startSaveNote(note))
        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        try {
            const { auth } = getState();
            await db.doc(`${auth.uid}/journal/notes/${id}`).delete();
            Swal.fire('Deleted', 'Note deleted', 'info');
            dispatch(deleteNote(id));
        } catch (error) {
            console.error(error);
        }
      
    }
}

const showUploadAlert = () => {
    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait',
        allowOutsideClick: false,
        willOpen: () => {
            Swal.showLoading()
        },
        showCloseButton: false,
        showCancelButton: false,
        showConfirmButton: false
    });
} 

export const addNewNote = (note) => ({
    type: types.notesAddnew,
    payload: {
        ...note
    }
});

export const setActiveNote = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
});

export const loadNotes = (notes) => ({
    type : types.notesLoad,
    payload: notes
});

export const refreshNote = (id, note) => ({
    type: types.notesUploaded,
    payload: {
        id,
        note: { id, ...note }
    }
});

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: {
        id
    }
});

export const notesLogout = () => ({
    type: types.notesCleaning,
});

