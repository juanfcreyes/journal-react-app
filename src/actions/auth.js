import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";
import { notesLogout } from "./notes";

export const startLogin = (email, password) => {
    return async (dispatch) => {
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password).then(({user}) => {
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
            dispatch(finishLoading());
            console.log('error', error);
        })
        
    }
}

export const startRegister = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(async ({user}) => {
            await user.updateProfile({displayName: name})
            dispatch(login(user.uid, user.displayName));
        }).catch((error) => {
            Swal.fire('Error', error.message, 'error');
        });
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        });
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(notesLogout()); 
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});