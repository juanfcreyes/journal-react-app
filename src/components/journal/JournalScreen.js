import React from 'react'
import { useSelector } from 'react-redux'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {
    const notes = useSelector(state => state.notes);
    return (
        <div className='journal__main-content animate__animated animate__fadeIn'>
            <Sidebar></Sidebar>
            <main>
                { notes.active ? 
                    <NoteScreen></NoteScreen> : 
                    <NothingSelected></NothingSelected>
                }
            </main>
        </div>
    )
}
