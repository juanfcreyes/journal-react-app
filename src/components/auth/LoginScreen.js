import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startLogin, startGoogleLogin } from '../../actions/auth'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    const ui = useSelector(state => state.ui);
    const dispatch = useDispatch();
        
    const [formValues, handleInputChange] = useForm(
        {
            email:'',
            password: ''
        });

    const {email, password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault(); 
        dispatch(startLogin(email, password));
    }

    const handleGoogleLogin = (e) => {
        e.preventDefault(); 
        dispatch(startGoogleLogin());
    }
    
    return (
        <>
            <h1 className='auth__title'> Login Screen</h1>
            <form onSubmit={handleLogin} className='animate__animated animate__fadeIn'>
                <input className='auth__input' autoComplete='off' type='text' placeholder='email' name='email' value={email} onChange={handleInputChange}/>
                <input className='auth__input' type='password' placeholder='password' name='password' value={password} onChange={handleInputChange}/>
                <button className='btn btn-primary btn-block mb-1 mt-1' type='submit' disabled={ui.loading}>Login</button>
                 <div className='auth__social-networks mb-1'>  
                    <p className='mb-1'>Login with social networks</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to='/auth/register' className='link'>
                    Create new account
                </Link>
            </form>
        </>
    )
}
