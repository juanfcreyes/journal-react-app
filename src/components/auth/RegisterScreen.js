import React from 'react'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { Link } from 'react-router-dom'
import { removeError, setError } from '../../actions/ui';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = () => {
    
    const ui = useSelector(state => state.ui); 
    const dispatch = useDispatch();

    const [formValues, handleInputChange] = useForm(
        {
            name: '',
            email:'',
            password1: '',
            password2: ''
        });

    const {name, email, password1, password2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(email, password1, name));
        }
    }

    const isFormValid = () => {

        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        }

        if (!validator.isEmail(email)) {
            dispatch(setError('Email not valid'));
            return false;
        }

        if (password1 !== password2 || password1.length < 5) {
            dispatch(setError('The password must have more than 5 characters and its confirmation must be the same'));
            return false;
        } 

        dispatch(removeError());

        return true;
    }

    return (
        <>
            <h1 className='auth__title'> Register</h1>
            <form onSubmit={handleRegister}  className='animate__animated animate__fadeIn'>
                {
                    ui.msgError &&
                    <div className='auth__alert-error'>        
                        { ui.msgError }
                    </div>
                }
                <input className='auth__input' autoComplete='off' type='text' placeholder='name' name='name' onChange={handleInputChange}  value={name}/>
                <input className='auth__input' autoComplete='off' type='text' placeholder='email' name='email' onChange={handleInputChange} value={email}/>
                <input className='auth__input' type='password' placeholder='password' name='password1' onChange={handleInputChange} value={password1}/>
                <input className='auth__input' type='password' placeholder='confirm password' name='password2' onChange={handleInputChange} value={password2}/>
                <button className='btn btn-primary btn-block mb-5 mt-1' type='submit'>Register</button>
                <Link to='/auth/login' className='link'>
                   have you already registered?
                </Link>
            </form>
        </>
    )
}
