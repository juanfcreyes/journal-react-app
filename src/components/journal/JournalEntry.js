import moment from 'moment'
import React from 'react'
import { useDispatch } from 'react-redux';
import { setActiveNote } from '../../actions/notes';

export const JournalEntry = ( {note} ) => {

    const dispatch = useDispatch();
    const momentDate = moment(note.date);

    const handleEntryClick = () => {
        dispatch(setActiveNote(note.id, note));
    } 

    return (
        <div className='journal__entry pointer animate__animated animate__fadeIn' onClick={handleEntryClick}>
            { note.url &&
                <div className='journal__entry-picture' style={{backgroundSize:'cover', backgroundImage:`url(${note.url})`}}></div>
            }
            <div className='journal__entry-body'>
                <p className='journal__entry-title'>{note.title}</p>
                <p className='journal__entry-content'> {note.body} </p>
            </div>
            <div className='jorunal__entry-date-box'>
                <span>{momentDate.format('dddd')}</span>
                <h4>{momentDate.format('D')}</h4>
            </div>
        </div>
    )
}
