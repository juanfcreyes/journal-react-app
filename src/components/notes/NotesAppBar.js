import React from 'react'
import moment from 'moment'
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const {active: note} = useSelector(state => state.notes);
    const dispatch = useDispatch();
    const momentDate = moment(note.date);
    
    const handleSave = () => {
        dispatch(startSaveNote(note));
    }

    const handlePictureUpload = () => {
       document.getElementById('file_input').click();
    }

    const handleFileChange = (e) => {
       const file = e.target.files[0];
       if (file) {
           dispatch(startUploading(file));
       }
    }

    return (
        <div className='notes__appbar'>
            <span>{momentDate.format('dddd DD MMMM YYYY')}</span>
            <input type='file' style={ {display: 'none' }} onChange={handleFileChange} id='file_input' name='file_input'></input>
            <div>
                <button className='btn' onClick={handlePictureUpload}>picture</button>
                <button className='btn' onClick={handleSave} >Save</button>
            </div>
        </div>
    )
}
