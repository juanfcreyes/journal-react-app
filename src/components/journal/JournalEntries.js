import React from 'react'
import { useSelector } from 'react-redux';
import { JournalEntry } from './JournalEntry';

export const JournalEntries = () => {
    const notesState = useSelector(state => state.notes);
    const entries = notesState.notes;    
    return (
        <div className='journal__entries animate__animated animate__fadeIn'>
            {
                entries.map((note) => (
                    <JournalEntry key={note.id} note={note} ></JournalEntry>
                ))
            }
        </div>
    )
}
