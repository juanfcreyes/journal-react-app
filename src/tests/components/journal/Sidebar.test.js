import { mount } from "enzyme";
import { Provider } from "react-redux";
import { Sidebar } from "../../../components/journal/Sidebar";

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";

jest.mock('../../../actions/auth', () => ({
    startLogout: jest.fn(),
}));


jest.mock('../../../actions/notes', () => ({
    startNewNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {

    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: null
    }
}

let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Prebas sobre el sidebar', () => {

    const wrapper = mount(
        <Provider store={store}>
            <Sidebar></Sidebar>
        </Provider>
    )
    
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar el logout', () => {
        wrapper.find('.btn').simulate('click');
        expect(startLogout).toHaveBeenCalled();
    });

    test('Debe de llamar el startNewNode', () => {
        wrapper.find('.journal__new-entry').simulate('click');
        expect(startNewNote).toHaveBeenCalled();
    });
    
});
