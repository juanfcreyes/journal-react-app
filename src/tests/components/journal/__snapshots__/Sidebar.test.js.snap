// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prebas sobre el sidebar Debe mostrarse correctamente 1`] = `
<aside
  className="journal_sidebar"
>
  <div
    className="journal_sidabar-narvar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span
        className="ml-1"
      >
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      new entry
    </p>
  </div>
  <div
    className="journal__entries animate__animated animate__fadeIn"
  />
</aside>
`;
